Sources/Algorithms/Compacted.swift

```diff
@@ -0,0 +1,174 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+public struct CompactedSequence<Base: Sequence, Element>: Sequence
+  where Base.Element == Element? {
+
+  @usableFromInline
+  let base: Base
+  
+  @inlinable
+  init(base: Base) {
+    self.base = base
+  }
+
+  public struct Iterator: IteratorProtocol {
+    @usableFromInline
+    var base: Base.Iterator
+    
+    @inlinable
+    init(base: Base.Iterator) {
+      self.base = base
+    }
+    
+    @inlinable
+    public mutating func next() -> Element? {
+      while let wrapped = base.next() {
+        guard let some = wrapped else { continue }
+        return some
+      }
+      return nil
+    }
+  }
+
+  @inlinable
+  public func makeIterator() -> Iterator {
+    return Iterator(base: base.makeIterator())
+  }
+}
+
+extension Sequence {
+  /// Complexity: O(1)
+  @inlinable
+  public func compacted<Unwrapped>() -> CompactedSequence<Self, Unwrapped>
+    where Element == Unwrapped? {
+    CompactedSequence(base: self)
+  }
+}
+
+public struct CompactedCollection<Base: Collection, Element>: Collection
+  where Base.Element == Element? {
+
+  @usableFromInline
+  let base: Base
+  
+  @usableFromInline
+  let _startIdx: Base.Index
+  
+  @inlinable
+  init(base: Base) {
+    self.base = base
+    self._startIdx = base.firstIndex(where: { $0 != nil }) ?? base.endIndex
+  }
+  
+  public struct Index {
+    @usableFromInline
+    let base: Base.Index
+    
+    @inlinable
+    init(base: Base.Index) {
+      self.base = base
+    }
+  }
+  
+  /// Complexity: O(1)
+  @inlinable
+  public var startIndex: Index { Index(base: _startIdx) }
+  
+  @inlinable
+  public var endIndex: Index { Index(base: base.endIndex) }
+  
+  @inlinable
+  public subscript(position: Index) -> Element {
+    base[position.base]!
+  }
+  
+  @inlinable
+  public func index(after i: Index) -> Index {
+    guard i.base != base.endIndex else {
+      return Index(base: base.endIndex)
+    }
+    
+    var idx = base.index(after: i.base)
+    while idx != base.endIndex {
+      guard base[idx] != nil else {
+        base.formIndex(after: &idx)
+        continue
+      }
+      return Index(base: idx)
+    }
+    return Index(base: base.endIndex)
```

Comment:
You could avoid a raw loop here by using `firstIndex(while:)` like you already do with `_startIdx`, but called on a slice instead (starting at `base.index(after: i.base)`). And similarly in `index(before:)`, but with `lastIndex(where:)`.

Reply:
I've adjusted, but actually I thought about it but just choose to use raw loop... but no strong feelings about it. Just fixed =]